src/chains/ethereum/ethereum/src/api.ts

```diff
@@ -2880,6 +2880,186 @@ export default class EthereumApi implements Api {
+        currentBlock.header.baseFeePerGas || Quantity.Zero;
+
+      const { gasUsed, gasLimit } = currentBlock.header;
+      if (gasUsed === gasLimit) {
+        gasUsedRatio[currentPosition] = 1;
+      } else {
+        gasUsedRatio[currentPosition] = Number(
+          `0.${(
+            (currentBlock.header.gasUsed.toBigInt() * PRECISION_BIG_INT) /
+            currentBlock.header.gasLimit.toBigInt()
+          )
+            .toString()
+            .padStart(PAD_PRECISION, "0")}`
+        );
+      }
+
+      if (reward) {
+        const transactions = currentBlock.getTransactions();
+
+        // If there are no transactions, all reward percentiles are 0.
+        if (transactions.length === 0) {
+          reward[currentPosition] = rewardPercentiles.map(() => {
+            return Quantity.Zero;
+          });
+        } else {
+          // For all transactions, effectiveGasReward = normalized fee per unit of gas
+          // earned by the miner regardless of transaction type
+          const baseFeePerGas = currentBlock.header.baseFeePerGas
+            ? currentBlock.header.baseFeePerGas.toBigInt()
+            : 0n;
+
+          const effectiveRewardAndGasUsed = (
+            await Promise.all(
+              transactions.map(async tx => {
+                let effectiveGasReward: bigint;
+                if ("maxPriorityFeePerGas" in tx) {
+                  effectiveGasReward =
+                    tx.maxFeePerGas.toBigInt() - baseFeePerGas;
+
+                  const maxPriorityFeePerGas =
+                    tx.maxPriorityFeePerGas.toBigInt();
+
+                  if (maxPriorityFeePerGas < effectiveGasReward) {
+                    effectiveGasReward = maxPriorityFeePerGas;
+                  }
+                } else {
+                  effectiveGasReward = tx.gasPrice.toBigInt() - baseFeePerGas;
+                }
+
+                const receipt = (
+                  await blockchain.transactionReceipts.get(tx.hash.toBuffer())
+                ).toJSON(currentBlock, tx, blockchain.common);
+                return {
+                  effectiveGasReward: effectiveGasReward,
+                  gasUsed: receipt.gasUsed ? receipt.gasUsed.toBigInt() : 0n
+                };
+              })
+            )
+          ).sort((a, b) => {
+            if (a.effectiveGasReward > b.effectiveGasReward) return 1;
+            if (a.effectiveGasReward < b.effectiveGasReward) return -1;
+            return 0;
+          });
+
+          // All of the block transactions are ordered, ascending, from least to greatest by
+          // the fee the tx paid per unit of gas. For each percentile of block gas consumed,
+          // what was the fee paid for the unit of gas at that percentile.
+          reward[currentPosition] = rewardPercentiles.map(percentile => {
+            let gasUsed = 0n;
+
+            const targetGas =
+              (currentBlock.header.gasUsed.toBigInt() *
+                BigInt(percentile * PRECISION_FLOAT)) /
```

Comment:
I don't _think_ there are actual floating point issues with `percentile * PRECISION_FLOAT` even though it could overflow `Number.MAX_SAFE_INTEGER` until we get to very large `percentile` values (way larger than `100`). BUT, should we convert percentile to `BigInt` and _then_ multiple by `PRECISION_FLOAT` (but as a bigint) just to be safe? :thinking:

Reply:
So this is one of those weirdies to me - `[0, 0.1234, 10, 50, 99, 100, 101, 200, 201.12309810294109]` is _technically_ in spec. 

One thing I noticed today testing against -> They now respond with `"invalid reward percentile: 110.000000"}` this must have recently changed or I completely missed it. 

I would like to respond in kind with an error to percentiles > 100. That _should_ mitigate the very valid insight on large percentile values and maintain our precision on arbitrarily pedantic length decimal percentiles.

Reply:
Oh, but it feels weird because `percentiles` considers `100.00` to be 100% and gasUsedRatio considers `1` to be 100%. It just feels odd even though I think they're appropriate for the name.

Reply:
Because `percentile` is now restricted to a range of `0 - 100` I think we're good. Recent geth updates now return an error for values below 0/above 100. 

Here are some notes I have on this: 
At 30M block gas and 100% usage the smallest range for `effectiveGasReward` is `.0007 || .07%`  (21,000 / 30M). So a `rewardPercentile` of `[50, 50.0007, 50.0014, 50.0021] _could_ return different values although it is highly unlikely now. The majority of the time these values will be functionally equivalent in ganache. 

If we cast to bigint up front, lopping off the decimal, we lose this and we differ from existing APIs that support it. E.g. we would throw on `[50, 50.0008,...]` and Infura/Alchemy, etc would not.

Reply:
Ok cool Sounds good to me. Do we throw now as well?

Reply:
:+1: 
Yes, we have a more specific error message thanks to @MicaiahReid https://github.com/trufflesuite/ganache/pull/3351#discussion_r982738245